python/ffsim/states/slater.py

```diff
@@ -0,0 +1,228 @@
+        )
+
+        if concatenate:
+            strings = concatenate_bitstrings(
+                strings_a,
+                strings_b,
+                BitstringType.INT,
+                length=len(orbs_a),
+            )
+            return convert_bitstring_type(
+                strings,
+                BitstringType.INT,
+                bitstring_type,
+                length=len(orbs_a) + len(orbs_b),
+            )
+
+        return convert_bitstring_type(
+            strings_a,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_a),
+        ), convert_bitstring_type(
+            strings_b,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _sample_slater_spinless(
+    rdm: np.ndarray,
+    nelec: int,
+    shots: int,
+    seed: np.random.Generator | int | None = None,
+) -> list[int]:
+    """Collect samples of electronic configurations from a Slater determinant.
+
+    The Slater determinant is defined by its one-body reduced density matrix (RDM).
+    The sampler uses a determinantal point process to auto-regressively produce
+    uncorrelated samples.
+
+    Args:
+        rdm: The one-body reduced density matrix that defines the Slater determinant.
+        shots: Number of samples to collect.
+        seed: A seed to initialize the pseudorandom number generator.
+            Should be a valid input to ``np.random.default_rng``.
+
+    Returns:
+        The sampled bitstrings in integer format.
+    """
+    norb, _ = rdm.shape
+
+    if nelec == 0:
+        return [0] * shots
+
+    if nelec == norb:
+        return [(1 << norb) - 1] * shots
+
+    rng = np.random.default_rng(seed)
+    samples = [_autoregressive_slater(rdm, norb, nelec, rng) for _ in range(shots)]
+    return [sum(1 << orb for orb in sample) for sample in samples]
+
+
+def _autoregressive_slater(
+    rdm: np.ndarray,
+    norb: int,
+    nelec: int,
+    seed: np.random.Generator | int | None = None,
+) -> list[int]:
+    """Autoregressively sample positions of particles for a Slater determinant wave
+    function using a determinantal point process.
+
+    Args:
+        rdm: A Numpy array with the one-body reduced density matrix.
+        norb: Number of orbitals.
+        nelec: Number of electrons.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+
+    Returns:
+        A Numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+    empty_orbitals = list(range(norb))
+    probs = np.diag(rdm).real / nelec
+    index = rng.choice(len(empty_orbitals), p=probs)
+    sample = [empty_orbitals[index]]
+    marginal = [probs[index]]
+    empty_orbitals.pop(index)
+    for _ in range(nelec - 1):
+        marginals = _generate_marginals(rdm, sample, empty_orbitals)
+        conditionals = marginals / marginal[-1]
```

Comment:
This line doesn't do anything because the next line normalizes the conditionals.